src/fs/file_system.rs

```diff
@@ -0,0 +1,289 @@
+//! Module for [FileSystem].
+
+use super::*;
+use crate::fs::dir_entry_iter::UefiDirectoryEntryIterator;
+use crate::fs::file_info::BoxedUefiFileInfo;
+use crate::fs::path::{Path, PathError};
+use alloc_api::string::{FromUtf8Error, String, ToString};
+use alloc_api::vec::Vec;
+use core::fmt;
+use core::fmt::{Debug, Formatter};
+use uefi::proto::media::file::FileType;
+use uefi::table::boot::ScopedProtocol;
+use uefi::{CString16};
+
+// #[derive(Debug, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+#[derive(Debug, Clone)]
+pub enum FileSystemError {
+    /// Can't open the root directory of the underlying volume.
+    CantOpenVolume,
+    /// The path is invalid because of the underlying [PathError].
+    IllegalPath(PathError),
+    /// The file or directory was not found in the underlying volume.
+    FileNotFound(String),
+    /// The path is existend but does not correspond to a directory when a directory was expected.
+    NotADirectory(String),
+    /// The path is existend but does not correspond to a file when a file was expected.
+    NotAFile(String),
+    /// Can't delete the file.
+    CantDeleteFile(String),
+    /// Error writing bytes.
+    WriteFailure,
+    /// Error flushing file.
+    FlushFailure,
+    /// Error reading file.
+    ReadFailure,
+    /// Can't parse file content as UTF-8.
+    Utf8Error(FromUtf8Error),
+    /// Could not open the given path.
+    OpenError(String),
+}
+
+impl From<PathError> for FileSystemError {
+    fn from(err: PathError) -> Self {
+        Self::IllegalPath(err)
+    }
+}
+
+/// Return type for public [FileSystem] operations.
+pub type FileSystemResult<T> = Result<T, FileSystemError>;
+
+/// Entry point into the file system abstraction described by the module description. Accessor to
```

Comment:
"described by the module description" sounds pleonastic 😆 let's say "the file system abstraction provided by this module"

Reply:
:D i change it